lib/models/StreamSplitter.js

```diff
@@ -0,0 +1,173 @@
+const crypto = require('crypto');
+const fs = require('fs');
+const path = require('path');
+const stream = require('stream');
+
+function randomString(size) {
+  return crypto.randomBytes(size).toString('base64url').slice(0, size);
+}
+
+class InMemoryStreamSplitter extends stream.EventEmitter {
+  constructor(stream, maxSize) {
+    super();
+
+    this.buffer = [];
+    this.stream = stream;
+    this.maxSize = maxSize;
+
+    this.partial = null;
+
+    this.stream.on('error', this.handleError.bind(this));
+    this.stream.on('data', this.handleData.bind(this));
+    this.stream.on('end', this.handleEnd.bind(this));
+  }
+
+  handleError(err) {
+    this.handleEnd();
+  }
+
+  handleEnd() {
+    if (this.partial) {
+      this.partial.stream.end();
+      this.partial = null;
+    }
+  }
+
+  handleData(data) {
+    if (!this.partial) {
+      this.partial = {
+        stream: new stream.PassThrough(),
+        bytes: 0,
+      };
+      this.emit('stream', this.partial.stream);
+    }
+
+    const overflow = this.partial.bytes + data.length - this.maxSize;
+    if (overflow > 0) {
+      this.handleData(data.splice(0, data.length - overflow));
+      this.handleData(data.splice(data.length - overflow, data.length));
+      return;
+    }
+
+    this.partial.bytes += data.length;
+    this.partial.stream.write(data);
+
+    if (overflow === 0) {
+      this.partial.stream.end();
+      this.partial = null;
+    }
+  }
+}
```

Comment:
```suggestion
```

Reply:
I recommend you release a 0.6 version prior to refactor as this might take some time to get to final version. I think tests for S3 should be made separately by refactoring some other tests and spliting them to general integration test that should work for all datastores and specific unit tests for each datastore.